pkg/v11/resource/tiller/tiller.go

```diff
@@ -0,0 +1,168 @@
+}
+
+// New creates a new configured tiller resource.
+func New(config Config) (*Resource, error) {
+	if reflect.DeepEqual(config.BaseClusterConfig, cluster.Config{}) {
+		return nil, microerror.Maskf(invalidConfigError, "%T.BaseClusterConfig must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.Tenant == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Tenant must not be empty", config)
+	}
+	if config.ToClusterGuestConfigFunc == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ToClusterGuestConfigFunc must not be empty", config)
+	}
+
+	r := &Resource{
+		baseClusterConfig:        config.BaseClusterConfig,
+		logger:                   config.Logger,
+		tenant:                   config.Tenant,
+		toClusterGuestConfigFunc: config.ToClusterGuestConfigFunc,
+	}
+
+	return r, nil
+}
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	clusterGuestConfig, err := r.toClusterGuestConfigFunc(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.ensureTillerInstalled(ctx, clusterGuestConfig)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	clusterGuestConfig, err := r.toClusterGuestConfigFunc(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.ensureTillerInstalled(ctx, clusterGuestConfig)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Resource) Name() string {
+	return Name
+}
+
+func (r *Resource) ensureTillerInstalled(ctx context.Context, clusterGuestConfig v1alpha1.ClusterGuestConfig) error {
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring tiller is installed")
+
+	clusterConfig, err := prepareClusterConfig(r.baseClusterConfig, clusterGuestConfig)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	tenantAPIDomain, err := key.APIDomain(clusterGuestConfig)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	tenantHelmClient, err := r.tenant.NewHelmClient(ctx, clusterConfig.ClusterID, tenantAPIDomain)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = tenantHelmClient.EnsureTillerInstalled(ctx)
+	if tenantcluster.IsTimeout(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "timeout fetching certificates")
+
+		// A timeout error here means that the cluster-operator certificate
+		// for the current guest cluster was not found. We can't continue
+		// without a Helm client. We will retry during the next execution, when
+		// the certificate might be available.
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+		resourcecanceledcontext.SetCanceled(ctx)
+
+		return nil
+	} else if helmclient.IsTillerInstallationFailed(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "Tiller installation failed")
```

Comment:
Isn't it better to just log the error? Why do we handle this in separation?

Reply:
I just moved the code from the chartoperator resource. I think it was so the resource was cancelled.

But yes it doesn't make sense since this is all the tiller resource does. I'll change it.